src/dev/HallSensor.cpp

```diff
@@ -0,0 +1,112 @@
+#include <EVT/io/GPIO.hpp>
+#include <EVT/utils/log.hpp>
+#include <HALf3/stm32f3xx_hal.h>
+#include <dev/HallSensor.hpp>
+#include <string>
+#include <valarray>
+
+namespace IO = EVT::core::IO;
+
+// TODO Tune these values
+constexpr uint32_t THRESHOLD = 5000;// Threshold for wheel speed
+
+namespace WSS::DEV {
+
+HallSensor::HallSensor(IO::GPIO& gpio, uint32_t wheelRadius, uint32_t numberOfMagnets) : gpio((IO::GPIO&) gpio), wheelRadius(wheelRadius), numberOfMagnets(numberOfMagnets) {
+    this->prevTime = 0;
+    this->lastInterval = 0;
+    this->state = WheelSpeedState::STOP;
+    this->magnetInLastRead = false;
+}
+
+void HallSensor::update() {
+    uint32_t now = HAL_GetTick();
+    uint32_t elapsedTime = now - prevTime;
+
+    // The magnet is detected if the previous gpio pin reading is the same as the current gpio
+    // pin reading
+    const bool magnetDetected = gpio.readPin() == MAGNET_DETECTED_STATE;
+    bool magnetDetectedNow = false;
+    if (!magnetWasDetected && magnetDetected != magnetInLastRead && magnetDetected) {
+        magnetWasDetected = true;
+        magnetDetectedNow = true;
+    } else if (magnetWasDetected && magnetDetected == magnetInLastRead) {
+        magnetWasDetected = false;
+    }
+    magnetInLastRead = magnetDetected;
```

Comment:
This whole section is a bit confusing, and I think there are a couple reasons. First, the variable names are too similar. Here, you use the term "detect" for two different things. `magnetDetected` and `magnetInLastRead` refer to the magnet being in front of the sensor. `magnetDetectedNow` and `magnetWasDetected` refer to the magnet having just begun passing by the sensor. I would recommend reaming the second set of variables to something like `rotationDetected` and `rotationInLastRead`. Regardless of your name choice, I'd also recommend adding a comment clarifying the purpose of each variable.

Second, I think the logic may be a bit overcomplicated. If I'm understanding correctly, the `else if` will only evaluate to true exactly one call after the `if` evaluates to true and sets `magnetWasDetected` to true. This is because when `magnetWasDetected` is true, I would expect `magnetDetected == magnetInLastRead` to also always evaluate to true.

Third, in general, I don't understand why we need `magnetWasDetected` at all. As I understand it, the goal of this if-else statement is just to detect a rising edge in the GPIO signal, which you should be able to get with something like this:
```
    if (!magnetInLastRead && magnetDetected) {
        magnetDetectedNow = true;
    }
```